server/observer.go

```diff
@@ -0,0 +1,199 @@
+// grpc interceptors that are provided by this interface.
+type GrpcObserverRegistry interface {
+	// UnaryInterceptor implements the grpc.UnaryServerInterceptor type to allow the
+	// registry to hook into unary grpc methods.
+	UnaryInterceptor(
+		context.Context,
+		interface{},
+		*grpc.UnaryServerInfo,
+		grpc.UnaryHandler) (interface{}, error)
+	// StreamInterceptor implements the grpc.StreamServerInterceptor type to allow the
+	// registry to hook into streaming grpc methods.
+	StreamInterceptor(
+		interface{},
+		grpc.ServerStream,
+		*grpc.StreamServerInfo,
+		grpc.StreamHandler) error
+	RegisterUnaryObserver(UnaryObserver)
+	RegisterStreamRequestObserver(StreamRequestObserver)
+	RegisterStreamResponseObserver(StreamResponseObserver)
+}
+
+// ShowcaseObserverRegistry returns the showcase specific observer registry.
+func ShowcaseObserverRegistry() GrpcObserverRegistry {
+	return &showcaseObserverRegistry{
+		uObservers:     map[string]UnaryObserver{},
+		sReqObservers:  map[string]StreamRequestObserver{},
+		sRespObservers: map[string]StreamResponseObserver{},
+	}
+}
+
+// showcaseObserverRegistry is an implementation of the ObserverRegistry. This registry
+// automatically handles DeleteTest requests and deletes the appropriate observers
+// for that request.
+type showcaseObserverRegistry struct {
+	mu             sync.Mutex
+	uObservers     map[string]UnaryObserver
+	sReqObservers  map[string]StreamRequestObserver
+	sRespObservers map[string]StreamResponseObserver
+}
+
+func (r *showcaseObserverRegistry) UnaryInterceptor(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler) (interface{}, error) {
+	resp, err := handler(ctx, req)
+
+	if info.FullMethod == "/google.showcase.v1alpha3.Testing/DeleteTest" {
+		err = r.deleteTestHandler(req)
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	for _, obs := range r.uObservers {
+		obs.ObserveUnary(ctx, req, resp, info, err)
+	}
+
+	return resp, err
+}
+
+func (r *showcaseObserverRegistry) deleteTestHandler(req interface{}) error {
+	deleteTestRequest, ok := req.(*pb.DeleteTestRequest)
+	if !ok {
+		return errors.New("Failed to delete the test")
+	}
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	_, uFound := r.uObservers[deleteTestRequest.GetName()]
+	_, sReqFound := r.sReqObservers[deleteTestRequest.GetName()]
+	_, sRespFound := r.sRespObservers[deleteTestRequest.GetName()]
+	if !(uFound || sReqFound || sRespFound) {
+		return fmt.Errorf("Could not find test: %s", deleteTestRequest.GetName())
+	}
+
+	delete(r.uObservers, deleteTestRequest.GetName())
+	delete(r.sReqObservers, deleteTestRequest.GetName())
+	delete(r.sRespObservers, deleteTestRequest.GetName())
+	return nil
+}
+
+type showcaseStream struct {
+	info     *grpc.StreamServerInfo
+	registry *showcaseObserverRegistry
+
+	grpc.ServerStream
+}
+
+func (s *showcaseStream) SendMsg(m interface{}) error {
+	s.registry.mu.Lock()
```

Comment:
Is the point of locking the stream message transfer with the same mutex as Observer mgmt to prevent the observer from being removed mid stream? If thats not the purpose, maybe we should have separate mutexes for stream transfer & observer mgmt.

Reply:
Correct, it is to prevent the observer from being removed mid stream.